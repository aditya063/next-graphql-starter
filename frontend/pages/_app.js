/* eslint-disable global-require */
/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable max-len */
/* eslint-disable react/no-danger */

import App from 'next/app';
import { ThemeProvider } from 'styled-components';

import theme from '../styles/theme.style';

export class MyApp extends App {
  componentDidMount() {
    if (process.env.NODE_ENV !== 'production') {
      const ReactDOM = require('react-dom');
      const ReactAxe = require('react-axe');

      // https://github.com/dequelabs/react-axe/issues/123
      const matches = node =>
        !(node.getAttribute('data-axe-reject') === 'true');

      ReactAxe(React, ReactDOM, 1000, {
        rules: [{ id: 'color-contrast', matches }]
      });
    }
  }

  render() {
    if (process.env.NODE_ENV === 'development') {
      console.log(
        '----------_app render----------',
        new Date().getMilliseconds()
      );

      // console.log('_app props: ', Object.keys(this.props));
      // console.log('_app props.apolloState: ', this.props.apolloState);

      // const cache = this.props.apollo.cache.extract();
      // console.log('apollo cache: ', cache);

      // pageProps includes data returned from getInitialProps
      // console.log('_app props.pageProps: ', Object.keys(this.props.pageProps));
    }

    const { Component, pageProps } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <Component {...pageProps} />
      </ThemeProvider>
    );
  }
}

// For FOUC issue, see _document.js
if (typeof window !== 'undefined') {
  window.addEventListener('DOMContentLoaded', () => {
    document.getElementById('holderStyle').remove();
  });
}

export default MyApp;
