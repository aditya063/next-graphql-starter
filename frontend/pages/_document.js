/* eslint-disable react/no-danger */
// https://github.com/zeit/next.js/tree/canary/examples/with-styled-components
import Document, { Head, Main, NextScript } from 'next/document';
import { ServerStyleSheet } from 'styled-components';

import GlobalStyle from '../styles/global.style';

class MyDocument extends Document {
  static async getInitialProps(ctx) {
    console.log('_document GIP', new Date().getMilliseconds());

    const sheet = new ServerStyleSheet();
    const originalRenderPage = ctx.renderPage;

    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props =>
            sheet.collectStyles(
              <>
                <GlobalStyle />
                <App {...props} />
              </>
            )
        });

      const initialProps = await Document.getInitialProps(ctx);

      return {
        ...initialProps,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        )
      };
    } finally {
      sheet.seal();
    }
  }

  render() {
    console.log('_document render', new Date().getMilliseconds());

    return (
      <html lang="en">
        <Head>
          <meta charSet="utf-8" />

          {/* SEO */}
          <meta name="description" content="" />

          {/* Responsive */}
          <meta
            name="viewport"
            content="
              width=device-width,
              height=device-height,
              initial-scale=1,
              minimum-scale=1,
              user-scalable=0,
              shrink-to-fit=no"
          />

          {/* Favicon */}
          <link
            rel="shortcut icon"
            type="image/x-icon"
            href="/static/favicon.ico"
          />

          {/* Webfonts */}
          <link
            rel="preload"
            href="/static/fonts/open-sans-v15-latin-regular.woff2"
            as="font"
            type="font/woff2"
            crossOrigin="anonymous"
          />
          <link
            rel="preload"
            href="/static/fonts/play-v10-latin-regular.woff2"
            as="font"
            type="font/woff2"
            crossOrigin="anonymous"
          />

          {/* Progressive Web App Manifest + Theme Color */}
          <link rel="manifest" href="/static/manifest.json" />
          <meta name="theme-color" content="#477CBF" />

          {/* Fixes flash of unstyled content for first load (Chromium bug)
            https://github.com/ant-design/ant-design/issues/16037
            Not only antd, but also any other style if you want to use ssr */}
          <style
            id="holderStyle"
            dangerouslySetInnerHTML={{
              __html: `
                *, *::before, *::after {
                transition: none !important;
                }
              `
            }}
          />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </html>
    );
  }
}

export default MyDocument;
