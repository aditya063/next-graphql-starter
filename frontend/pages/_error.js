import PropTypes from 'prop-types';
import Link from 'next/link';

class ErrorPage extends React.PureComponent {
  static getInitialProps({ res, err }) {
    const statusCode = res ? res.statusCode : err ? err.statusCode : null;
    return { statusCode };
  }

  render() {
    let response;
    switch (this.props.statusCode) {
      case undefined:
        response = (
          <>
            <p>Unknown error, no status code to report.</p>{' '}
            <Link href="/">
              <a>Home</a>
            </Link>
          </>
        );
        break;
      case 404:
        response = (
          <>
            <h1>Page Not Found</h1>
            <Link href="/">
              <a>Home</a>
            </Link>
          </>
        );
        break;
      case 500:
        response = (
          <>
            <h1>Internal Server Error</h1>
            <Link href="/">
              <a>Home</a>
            </Link>
          </>
        );
        break;
      default:
        response = (
          <>
            <h1>HTTP {this.props.statusCode} Error</h1>
            <Link href="/">
              <a>Home</a>
            </Link>
          </>
        );
        break;
    }

    return response;
  }
}

ErrorPage.defaultProps = {
  statusCode: null
};

ErrorPage.propTypes = {
  statusCode: PropTypes.number
};

export default ErrorPage;
