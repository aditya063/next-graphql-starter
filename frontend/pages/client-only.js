
import { withApollo } from '../graphql/apollo';

const ClientOnlyPage = () => (
  <div>

    This example shows how to disable apollos query fetching on the server.
    If you
    {' '}
    <a href="/client-only">reload</a>
    {' '}
    this page, you will see a loader since Apollo
    didnt fetch any data on the server. This allows
    {' '}
    <a
      href="https://nextjs.org/blog/next-9#automatic-static-optimization"
    >
      automatic static optimization
    </a>
    .

    <article>
      <h1>The Idea Behind This Example</h1>
      <p>
        <a href="https://www.apollographql.com/client/">Apollo</a>
        is a GraphQL
        client that allows you to easily query the exact data you need from a
        GraphQL server. In addition to fetching and mutating data, Apollo
        analyzes your queries and their results to construct a client-side cache
        of your data, which is kept up to date as further queries and mutations
        are run, fetching more results from the server.
      </p>
      <p>
        In this simple example, we integrate Apollo seamlessly with
        {' '}
        <a href="https://github.com/zeit/next.js">Next</a>
        {' '}
        by wrapping our Page
                component inside a
        {' '}
        <a href="https://facebook.github.io/react/docs/higher-order-components.html">
          higher-order component (HOC)
        </a>
        . Using the HOC pattern were able to pass down a central store of query
        result data created by Apollo into our React component hierarchy defined
        inside a page of our Next application.
      </p>
      <p>
        On initial page load, while on the server and inside getInitialProps, we
        invoke the Apollo method,
        {' '}
        <a href="https://www.apollographql.com/docs/react/api/react-ssr/#getdatafromtree">
          getDataFromTree
        </a>
        . This method returns a promise; at the point in which the promise
        resolves, our Apollo Client store is completely initialized.
      </p>
      <p>
        This example relies on
        {' '}
        <a href="http://graph.cool">graph.cool</a>
        {' '}
        for
                its GraphQL backend.
      </p>
    </article>
  </div>
);

export default withApollo(ClientOnlyPage, {
  // Disable apollo ssr fetching in favour of automatic static optimization
  ssr: false
});
