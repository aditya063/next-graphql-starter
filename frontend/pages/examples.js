import Head from 'next/head';
import Examples from '../components/Examples/Examples';

class ExamplePage extends React.PureComponent {
  render() {
    return (
      <>
        <Head>
          <title>next-graphql-starter | Examples</title>
        </Head>
        <Examples />
      </>
    );
  }
}

export default ExamplePage;
