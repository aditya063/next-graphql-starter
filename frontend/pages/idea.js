import PropTypes from 'prop-types';

import Head from '../components/Head/Head';
import LayoutMain from '../components/LayoutMain/LayoutMain';
import HeaderDetail from '../components/HeaderDetail/HeaderDetail';
import IdeaDetail from '../components/IdeaDetail/IdeaDetail';
import { withApollo } from '../graphql/withApollo';
import authenticate from '../utils/authenticate';

const IdeaPage = props => (
  <>
    <Head title="Idea Detail" />

    <LayoutMain
      header={<HeaderDetail ideaId={props.ideaId} />}
      content={<IdeaDetail ideaId={props.ideaId} />}
    />
  </>
);

IdeaPage.getInitialProps = ctx => {
  const { req, res, pathname, query } = ctx;

  if (req && res && pathname) {
    authenticate(req, res, pathname);
  }

  const ideaId = query.id;

  return { ideaId };
};

IdeaPage.propTypes = {
  ideaId: PropTypes.string.isRequired
};

export default withApollo(React.memo(IdeaPage));
