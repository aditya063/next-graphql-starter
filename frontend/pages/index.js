import { withApollo } from '../graphql/withApollo';
import Page from '../containers/Page/Page';
import Head from '../containers/Head/Head';
import LayoutMain from '../containers/LayoutMain/LayoutMain';
import HeaderMain from '../containers/HeaderMain/HeaderMain';
import Ideas from '../components/Ideas/Ideas';

const IndexPage = React.memo(() => (
  <Page>
    <Head title="Home" />

    <LayoutMain header={<HeaderMain />} content={<Ideas />} />
  </Page>
));

export default withApollo(IndexPage);
