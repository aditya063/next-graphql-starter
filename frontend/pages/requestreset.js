import Head from '../components/Head/Head';
import RequestReset from '../components/RequestReset/RequestReset';
import { redirect } from '../utils/redirect';
import { isLoggedIn } from '../utils/isLoggedIn';

const RequestResetPage = React.memo(() => (
  <>
    <Head title="Request Reset" />

    <RequestReset />
  </>
));

RequestResetPage.getInitialProps = async ctx => {
  const me = await isLoggedIn(ctx.apolloClient);

  if (me) redirect(ctx, '/');
};

export default RequestResetPage;
