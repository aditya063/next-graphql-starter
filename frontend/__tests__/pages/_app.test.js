import React from 'react';
import { shallow } from 'enzyme';
import { MyApp } from '../../pages/_app';

describe('_app', () => {
  let mockProps;
  let wrapper;

  beforeEach(() => {
    jest.resetAllMocks();
    mockProps = {};
    wrapper = shallow(<MyApp {...mockProps} />, {
      disableLifecycleMethods: true
    });
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
