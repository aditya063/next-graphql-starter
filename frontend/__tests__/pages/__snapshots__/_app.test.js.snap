// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_app matches snapshot 1`] = `
<Container>
  <ApolloProvider>
    <ThemeProvider
      theme={
        Object {
          "color": Object {
            "black": "#393939",
            "blue": "#ACE",
            "green": "limegreen",
            "orange": "tomato",
            "purple": "blueviolet",
            "yellow_bright": "#FFFF99",
            "yellow_light": "#fffacd",
          },
        }
      }
    >
      <Component>
        <Component />
      </Component>
    </ThemeProvider>
  </ApolloProvider>
</Container>
`;
